### PR TITLE
Netdata: Enable Netdata cloud integration

### LINKED ISSUE #9061: [macOS / Homebrew] This agent was built without the dependencies for Cloud and cannot be claimed
##### Bug report summary
Installed the latest version with brew upgrade netdata
Then tried to claim the node using the following:
`sudo netdata-claim.sh -token=tokenkey -rooms=room-key -url=https://app.netdata.cloud`
The following message is returned:

> This agent was built without the dependencies for Cloud and cannot be claimed

##### OS / Environment
10.15.4 (19E287)

Place the output from the command in the code section below.  
 -->
```
This agent was built without the dependencies for Cloud and cannot be claimed
```

##### Steps To Reproduce

1. ...
2. ...

### PR DESCRIPTION
**Pull request by rickard-von-essen**
Netdata cloud integration was missing becuse it requires two statically linked libraries,
libwebsockets and libmosquitto.

Fixes netdata/netdata#9061

-----

- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

### CODE REVIEW

**Review by prologic — APPROVED** [review]
LGTM (_aside from the CI failure_) this lines up with how our installer builds the `netdata` binary with Netdata Cloud support! 🙇‍♂️

File: Formula/netdata.rb
Comment (SMillerDev) [inline Formula/netdata.rb:52]: These should really be made to work with the homebrew formula. We don't want to vendor anything we have a formula for.

Comment (Ferroin) [reply Formula/netdata.rb:52]: libmosquito has to be vendored at the moment, we need a custom patch on top of upstream for it to work. We're working on getting that upstream, but I don't know of any ETA on that.

LWS, OTOH, could theoretically be handled with a system copy (provided the system copy is built with SOCKS5 support), but our build system currently does not support that approach.

We have an internal issue tracking support for this here: https://github.com/netdata/netdata/issues/8961

**Comment by rickard-von-essen** [comment]
Some one needs to kick CI.

File: Formula/netdata.rb
Comment (SMillerDev) [reply Formula/netdata.rb:52]: In that case we should keep cloud integration disabled until it can meet homebrew requirements.

Comment (rickard-von-essen) [reply Formula/netdata.rb:52]: > We don't want to vendor anything we have a formula for.

I understand this argument I think it's a bit arbitrary since for Python and Ruby formulas this is exactly what is done (in contrary with how the "native" packaging/dependency management work).

I would say that in this case it should be allowed since this statically links these dependencies into the binary and Netdata is very actively maintained will likely resolve these issues in the future and then remove the vendoring.

Also while checking the guidelines for acceptable formals etc I couldn't find any mentioning that this is explicitly disallowed.

Comment (SMillerDev) [reply Formula/netdata.rb:52]: > I understand this argument I think it's a bit arbitrary since for Python and Ruby formulas this is exactly what is done (in contrary with how the "native" packaging/dependency management work).

I'm not sure what you mean but nothing here is vendored: https://github.com/Homebrew/homebrew-core/blob/master/Formula/ruby.rb

> I would say that in this case it should be allowed since this statically links these dependencies into the binary and Netdata is very actively maintained will likely resolve these issues in the future and then remove the vendoring.

Looking at the open issue that has been parked for months doesn't fill me with confidence that this will be resolved any time soon.

> Also while checking the guidelines for acceptable formals etc I couldn't find any mentioning that this is explicitly disallowed.

You're right, I'll make a pull request to improve that.

Comment (rickard-von-essen) [reply Formula/netdata.rb:52]: > I'm not sure what you mean but nothing here is vendored: https://github.com/Homebrew/homebrew-core/blob/master/Formula/ruby.rb

Like this https://github.com/Homebrew/homebrew-core/blob/master/Formula/awscli.rb#L28-L30

> Looking at the open issue that has been parked for months doesn't fill me with confidence that this will be resolved any time soon.

I specifically avoided the word soon. :wink: 

> You're right, I'll make a pull request to improve that.

:+1:

Comment (SMillerDev) [reply Formula/netdata.rb:52]: > Like this https://github.com/Homebrew/homebrew-core/blob/master/Formula/awscli.rb#L28-L30

We vendor python dependencies cause pip...is difficult to work with as a package manager. If python libraries worked the same way system libraries do we could remove a lot of duplication.

Comment (prologic) [reply Formula/netdata.rb:52]: I'm just going to say this once. I _really_ don't understand why in general (_this isn't aimed at Homebrew core at all!_) we get a bit zealous with packaging and vendoring. A package manager is a tool, we use it as such to "package software". However that is accomplished is really not the point here there are _often_ **very good** reasons to vendor a dependency or pull in sources from other sources (_depending on licensing/needs_).

As my colleague @Ferroin points out we _cannot_ current ship a Netdata Agent built with cloud support with _any_ system copy of the libwebsockets until we get the patches we had to make upstream (_currently no EAT on this_).

If we can't get this PR into Homebrew core, then we will remove it and offer Netdata as a Tap instead at https://github.com/netdata.

Comment (SMillerDev) [reply Formula/netdata.rb:52]: I understand that sometimes there are patches that need to be done and homebrew will happily accept patches that make software work better for everyone. But if someone finds a CVE in libwebsockets we would really rather patch that in one place for all formula, rather than going through all of them and reminding authors that they need to update their vendored dependencies. If you can patch the current libwebsockets formula with some PR you made upstream that isn't released yet, that would be the best way forward IYAM.

As for removing the formula, we won't do that. This is open source software and homebrew is free to distribute it.
https://docs.brew.sh/How-to-Create-and-Maintain-a-Tap#official-vendor-taps

Comment (prologic) [reply Formula/netdata.rb:52]: @SMillerDev Thanks for your understanding and empathy. We _are_ trying out best to get our patches upstream. In the meantime we are a bit stuck.. I _get_ the whole "security" argument, however the argument can be made that it is just as much as "our" responsibility to patch any source, vendor'd libraries, etc as a company as it is for maintainers of open-source packages. In any case, I'll let you guys decide what you want to do here, We _cannot_ easily get the required pathces upstream in any reasonable timeframe I'm afraid, so bare that in mind.

Comment (SMillerDev) [reply Formula/netdata.rb:52]: I think it's both our responsibilities indeed. You have the responsibility to your customers and homebrew has the responsibility to its users.

It's it's okay if you can't get those patches upstreamed, but in that case I'd like to hold off on adding the cloud feature to the homebrew formula until you can.

Comment (SMillerDev) [reply Formula/netdata.rb:52]: Documentation is now available mentioning our hesitance to include vendored dependencies: https://docs.brew.sh/Acceptable-Formulae#stuff-that-requires-vendored-versions-of-homebrew-formulae

Should I close this PR so you can make a new one when it meets the requirements for properly using system dependencies?

Comment (prologic) [reply Formula/netdata.rb:52]: Thanks for your feedback and time taken to document Homebrew's position and your understanding. Yes I _think_ we can close this PR for the time being as I don't believe we can realistically get these patches upstream right now. Thanks!

**Comment by Ferroin** [comment]
Probably of interest to anybody who was following this, Netdata now supports building against a system copy of LWS (and also supports LWS 4.1 instead of just LWS 3.x, see https://github.com/netdata/netdata/pull/9984 for more info).

Unfortunately, we still need a static copy of mosquitto with custom patches, but there are plans in the works to completely eliminate both mosquitto and LWS as dependencies.

**Pull request closed without merging**
